javascript/ql/src/semmle/javascript/security/dataflow/ZipSlip.qll

```diff
@@ -0,0 +1,124 @@
+
+    override predicate isSource(DataFlow::Node source) { source instanceof Source }
+
+    override predicate isSink(DataFlow::Node sink) { sink instanceof Sink }
+
+    override predicate isSanitizerGuard(TaintTracking::SanitizerGuardNode nd) {
+      nd instanceof SanitizerGuard
+    }
+  }
+
+  /**
+   * Holds if `node1` flows to `node2` in one step by virtue of
+   * `node2` being of the form `.pipe(node1)`. The reason this flow
+   * exists is that `.pipe` returns its argument to make chained
+   * stream operations work.
+   */
+  predicate pipeStep(DataFlow::Node node1, DataFlow::MethodCallNode node2) {
+      node2.getMethodName() = "pipe" and
+      node1 = node2.getArgument(0)
+  }
+
+  /**
+   * Holds if `node1` flows to `node2` in one step including the assumption that
+   * `x` flows to `.pipe(x)`
+   */
+  predicate stepsThroughPipe(DataFlow::Node node1, DataFlow::Node node2) {
+    DataFlow::localFlowStep(node1, node2) or pipeStep(node1, node2)
+  }
+
+  /**
+   * Holds if `node1` flows to `node2` including the assumption that
+   * `x` flows to `.pipe(x)`
+   */
+  predicate flowsThroughPipe(DataFlow::Node node1, DataFlow::Node node2) {
+    stepsThroughPipe*(node1, node2)
+  }
+
+  /**
+   * An access to the filepath of an entry of a zipfile being extracted
+   * by npm module `unzip`. For example, in
+   * ```javascript
+   * const unzip = require('unzip');
+   *
+   * fs.createReadStream('archive.zip')
+   *   .pipe(unzip.Parse())
+   *   .on('entry', entry => {
+   *      const path = entry.path;
+   *   });
+   * ```
+   * there is an `UnzipEntrySource` node corresponding to
+   * the expression `entry.path`.
+   */
+  class UnzipEntrySource extends Source {
+    UnzipEntrySource() {
+      exists(DataFlow::SourceNode parsed |
+        flowsThroughPipe(DataFlow::moduleImport("unzip").getAMemberCall("Parse"), parsed)
+        and
+        this = parsed.getAMemberCall("on").getCallback(1).getParameter(0).getAPropertyRead("path")
+      )
+    }
+  }
+
+  /**
+   * A sink that is the path that a createWriteStream gets created at.
+   * This is not covered by FileSystemWriteSink, because it is
+   * required that a write actually takes place to the stream.
+   * However, we want to consider even the bare createWriteStream to
+   * be a zipslip vulnerability since it may truncate an existing file.
+   */
+  class CreateWriteStreamSink extends Sink {
+    CreateWriteStreamSink() {
+      this = DataFlow::moduleImport("fs").getAMemberCall("createWriteStream").getArgument(0)
+    }
+  }
+
+  /** A sink that is a file path that gets written to. */
+  class FileSystemWriteSink extends Sink {
+    FileSystemWriteSink() { exists(FileSystemWriteAccess fsw | fsw.getAPathArgument() = this) }
+  }
+
+  /**
+   * Gets a string which suffices to search for to ensure that a
+   * filepath will not refer to parent directories.
+   */
+  string getAParentDirName() { result = any(string s | s = ".." or s = "../") }
```

Comment:
equivalent:

```
result = ".." or result = "../"
```
But in this simple case, you may as well just inline this getter and save a docstring.

---

I do wonder though: why not `"/.."` or `"/../"`? Perhaps it is time to use `s.regexpMatch("/?\\.\\./?")`... (but if you do that, then you need to inline explicitly or use the predicate annotation `bindingset[result]`)

Reply:
It was @asger-semmle's suggestion to add `"../"` in the first place. On second thought, I'm a little unsure why it's a good idea to allow a *larger* string `s` in a position where we are saying that checking for noninclusion of `s` counts as santitizing. The library user *checking* for `"/.."` or `"/../"` would be unsound, as they'd miss malicious paths `../sneakyFile` that go back only one directory.

Reply:
hmm, perhaps you should just use the similar sanitizer `StrongPathCheck` in `TaintedPath.qll`. That has been tested a lot.

In general, we are moving towards identifying insufficient/suspicious sanitizers with a separate set of queries, see `IncompleteUrlSubstringSanitization.ql`.

Reply:
I think the current sanitizer is fine for now. Zip files cannot contain absolute paths nor Windows paths, so I'd expect sanitization to be a bit simpler here than in TaintedPath.